pkg/devices/common.go

```diff
@@ -13,350 +16,474 @@ import (
+		} else {
+			log.Debugf("Key present")
+			// See if this fits on the end of the slice.
+			//  If so, update the ModbusBulkRead RegisterCount.
+			//  If not, create a new ModbusBulkRead.
+			reads := bulkReadMap[key]
+			lastRead := reads[len(reads)-1]
+			startRegister := lastRead.StartRegister
+			log.Debugf("startRegister: 0x%0x", startRegister)
+			newRegisterCount := deviceDataAddress + deviceDataWidth - startRegister
+
+			if newRegisterCount <= key.MaximumRegisterCount {
+				log.Debugf("read fits in existing. newRegisterCount: %v", newRegisterCount)
+				lastRead.RegisterCount = newRegisterCount
+				lastRead.Devices = append(lastRead.Devices, device)
+			} else {
+				// Add a new read.
+				log.Debugf("read does not fit in existing. newRegisterCount: %v", newRegisterCount)
+				modbusBulkRead, err := NewModbusBulkRead(device, deviceDataAddress, deviceDataWidth, isCoil)
+				if err != nil {
+					return nil, keyOrder, err
+				}
+				log.Debugf("modbusBulkRead: %#v", modbusBulkRead)
+				bulkReadMap[key] = append(bulkReadMap[key], modbusBulkRead)
+			}
+		}
+	} // For each device.
+	return bulkReadMap, keyOrder, nil
+}
 
-	log.WithFields(log.Fields{
-		"coilByte":  coilByte,
-		"mask":      mask,
-		"coilState": coilState,
-	}).Debug("calculating coil state")
+// MapBulkReadData maps the data read over modbus to the device read contexts.
+func MapBulkReadData(bulkReadMap map[ModbusBulkReadKey][]*ModbusBulkRead, keyOrder []ModbusBulkReadKey) (
+	readContexts []*sdk.ReadContext, err error) {
+	// This map tells us if we have already created a read context for this
+	// device and output. We can hit the same device and output more than once in
+	// this loop when there are multiple modbus reads for a single device (more
+	// than the maximum number ofregister addresses)
+	accountedFor := make(map[*sdk.Device]*output.Output)
+
+	for a := 0; a < len(keyOrder); a++ {
+		k := keyOrder[a]
+		v := bulkReadMap[k]
+		for h := 0; h < len(v); h++ { // for each read
+			read := v[h]
+			devices := read.Devices
+
+			for i := 0; i < len(devices); i++ {
+				device := devices[i]
+
+				theOutput := output.Get(device.Output)
+				readings := []*output.Reading{}
+
+				// Have we accounted for this device and output yet?
+				// This can happen when multiple reads are required for a single ModbusBulkReadKey.
+				_, keyPresent := accountedFor[device]
+				inMap := false
+				if keyPresent {
+					// Device is there. Is the output there?
+					if accountedFor[device] == theOutput {
+						inMap = true
+						break // for
+					}
+					//}
+					if inMap {
+						log.Debugf("device[output] already accounted for: device %p, output %p", device, theOutput)
+						continue // next output
+					}
+				}
+
+				// Get address and width.
+				var deviceData config.ModbusDeviceData
+				err := mapstructure.Decode(device.Data, &deviceData)
+				if err != nil { // This is not a configuration issue. Device may not have responded.
```

Comment:
I'm not sure what this comment means; wouldn't it be a configuration issue here, since this isn't dealing with device response, just decoding the device Data field into a struct